### DISCUSSION TITLE
Some questions and suggestions

### DISCUSSION THREAD

**Discussion by jcandres** [OctoD/godot-gameplay-systems#32] [Ideas]
Wow, what a great project. I have been making my own GAS adaptation for a while, and just saw this addon.

I've been playing with the examples and trying to understand how everything works. I'm very impressed do far. Here's some issues I've found while exploring GGS, in particular the abilities and attributes system: 

- Can you add/remove tags with an Effect? If not, how can you implement status effects? It would be handy to simply specify what tags are added/removed each time the effect is applied, and also what others change when the effect is removed.
- Is there a reason why `AbilityContainer` is not the main component that holds attributes and handles effect application? For me one of the selling points of this particular system is that everything is handled from a single component. In particular I believe that effects should be applied and removed via  `AbilityContainer` (that checks conditional application tags, and also be affected by the effect's tags) and not directly by `GameplayAttributeMap`. 
- Do you have any plans of adding turn-based options? For example, you could  make the ability cooldowns and effect durations use a callback for their timers. This way the callback can be invoked manually in turn-based games without too much fuss. 
- Is there a built-in way to do custom effect calculations? I believe right now you specify a min-max range, hardcoded in the effect. It would be great if instead you could specify a class that uses the source and target attributes involved in the ability to do some custom calculation.

Anyway, thanks for the addon and keep up the good work!

**Comment by OctoD** [comment]
Hi, thank you for the questions!

> Can you add/remove tags with an Effect? If not, how can you implement status effects? It would be handy to simply specify what tags are added/removed each time the effect is applied, and also what others change when the effect is removed.

It will be done in the next releases (maybe in a week, I am currently working on camera shake effect nodes for both 2D and 3D)

> Is there a reason why AbilityContainer is not the main component that holds attributes and handles effect application? For me one of the selling points of this particular system is that everything is handled from a single component. In particular I believe that effects should be applied and removed via AbilityContainer (that checks conditional application tags, and also be affected by the effect's tags) and not directly by GameplayAttributeMap.

I decided to split them because, in certain circumstances, having a big node which handles everything would be an overkill. 

Some games just need attributes and not abilities (like a RTS, or maybe in some other games a lot of enemies have only the health attribute), so I decided to separate the things. They can talk to each other though by binding their own signals.

> Do you have any plans of adding turn-based options? For example, you could make the ability cooldowns and effect durations use a callback for their timers. This way the callback can be invoked manually in turn-based games without too much fuss.

It will come with a turn based update, stay tuned :D 

> Is there a built-in way to do custom effect calculations? I believe right now you specify a min-max range, hardcoded in the effect. It would be great if instead you could specify a class that uses the source and target attributes involved in the ability to do some custom calculation.

If you create an effect as a scene, you can edit stuff from the editor. I just hardcoded the examples because I am lazy, but I will certanly add a more "godot" way of making them!

Thank you again, do not hesitate to ask, suggest or contribute!

**Reply by jcandres** [reply]
Thanks for the quick reply!


> I decided to split them because, in certain circumstances, having a big node which handles everything would be an overkill.

That makes sense and I appreciate how modular everything is.

> If you create an effect as a scene, you can edit stuff from the editor. 

To clarify, what I mean is that `AttributeEffect.get_current_value()` will return a random value. I find this a bit restricting, since it means that you have to calculate the min and max values beforehand, at a point where the source and target of the ability are known -- where would you calculate this? In the `Ability` itself?
I don't think you need a "godot way" for this, simply passing the `ActivationEvent` along with the effect could be used to do some calculation when it's applied to a target.

Another question: is there an easy way to implement a dynamic 'chain' of abilities? For example an ability that makes the user move to a point and then swing a sword/shoot a bow/open a chest. At the moment I'm doing this with a monolithic, multi-stage ability, but perhaps there's a better way. 

> It will come with a turn based update, stay tuned :D

Awesome!


Thanks again for all the hard work.

**Comment by OctoD** [comment]
> To clarify, what I mean is that AttributeEffect.get_current_value() will return a random value. I find this a bit restricting, since it means that you have to calculate the min and max values beforehand, at a point where the source and target of the ability are known -- where would you calculate this? In the Ability itself?
I don't think you need a "godot way" for this, simply passing the ActivationEvent along with the effect could be used to do some calculation when it's applied to a target.

The value is random only if you set a different minimum and a different maximum value applied, if both are the same the value will be constant. May I ask you for a real world example please? I could add anything based on devs needs!


> Another question: is there an easy way to implement a dynamic 'chain' of abilities? For example an ability that makes the user move to a point and then swing a sword/shoot a bow/open a chest. At the moment I'm doing this with a monolithic, multi-stage ability, but perhaps there's a better way.

Starting from the "open a chest" example, I would suggest to use the interactables nodes present in this addon, they are made for these kind of tasks and should make your life easier!

Concerning the multistage abilities, at this moment there is not a smooth way to implement. I would suggest to implement different abilities which can be activated automatically when certain the AbilityContainer node has some tags. You will have:

- Ability1, which adds at it's end the tag "ability1.ended" 
- Abiltiy2 which starts only if "ability1.ended" is there and adds another tag "ability2.ended" and so on

I hope it helped, otherwise I will study a good approach for this use case, thank you!

**Comment by BlazerGamer** [comment]
Good afternoon friend! I just added something to the Item.gd script, since I'm new to this godot engine, I added a new function called rest_quantity(value) since I didn't know how to subtract the amount that the item's stack had. item hahaha, I would also like to be able to help you with the visual part of the inventory, since it seems to me the most flexible I have seen

**Reply by OctoD** [reply]
Hi, thank you for writing. May I ask you what are you trying to accomplish? I did not grasp the gist of the question. Thank you!
